pkg/lisafs/connection.go

```diff
@@ -81,11 +81,13 @@ type Connection struct {
 	fds map[FDID]genericFD
 	// nextFDID is the next available FDID. It is protected by fdsMu.
 	nextFDID FDID
+	// the user id of the sandboxed application, specified in the Spec for the container.
+	appUid uint32
```

Comment:
As discussed, the app UID can change. This should not be a property of the lisafs connection, which persists through the lifetime of the container. gVisor has many users and we need to think about all use-cases. You need to pass the application thread's UID in the lisafs message. For example, see how mkdir is implemented:

- The gofer client fetches the application threads' credentials and makes lisafs RPC with that: https://github.com/google/gvisor/blob/44562d85cf0642fb83ec30f27b450d2fcaa9a722/pkg/sentry/fsimpl/gofer/filesystem.go#L831-L842.
- In your case, you would just need `rp.Credentials().EffectiveKUID`.
- You need to pass UID as part of Connect RPC. However, for backward compatibility, we can not just update the RPC message. You will need to create a new variant of the Connect RPC (maybe `ConnectWithUID` or `Connect2`) which will define a new message with the UID defined this time. See https://github.com/google/gvisor/commit/d5002c6adc315bb0efa958917e6b2fb13c8a434f as an example commit which implements a new RPC in lisafs.
- You will need to update [`ClientFD.Connect()`](https://github.com/google/gvisor/blob/44562d85cf0642fb83ec30f27b450d2fcaa9a722/pkg/lisafs/client_file.go#L470) to also take in a UID parameter. Then have it check if the new variant of Connect is supported by the gofer. For example, [like this](https://github.com/google/gvisor/blob/44562d85cf0642fb83ec30f27b450d2fcaa9a722/pkg/lisafs/client_file.go#L417-L420). If it is supported, then it should make the newer version of the Connect RPC with the UID. If not, then it will fallback to the older one.
- In the gofer, implement the new Connect RPC to temporarily switch the UID and just call the handler of the old Connect RPC.

Reply:
Thank you so much for the detailed instructions! I appreciate this a lot.

I will follow your suggestions. In the mean time I may still push many commits, but please only review this PR when I ping you & request your review! thanks.

Reply:
@ayushr2 I made sure all the unit tests pass and this PR is reviewable now. While I try to come up with some integration test case, could you take a look at the current PR? Thank you so much!